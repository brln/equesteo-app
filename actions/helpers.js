import BackgroundGeolocation from 'react-native-mauron85-background-geolocation'
import firebase from 'react-native-firebase'

import {
  dismissError,
  errorOccurred,
  loadCurrentRideState,
  loadLocalState,
  localDataLoaded,
  saveUserID,
  setAwaitingPasswordChange,
  setDoingInitialLoad,
} from './standard'

import {
  getFCMToken,
  startListeningFCMTokenRefresh,
  startListeningFCM,
  setDistributionOnServer,
  switchRoot,
} from './functional'
import { logError, logInfo } from '../helpers'
import { FEED } from '../screens'
import { LocalStorage, PouchCouch } from '../services'

import { setUserContext } from "../services/Sentry"

export function loginAndSync(loginFunc, loginArgs, dispatch) {
  loginFunc(...loginArgs).then(resp => {
    // @TODO: figure out why followers have _id here!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1
    const userID = resp.id
    const following = resp.following
    const followers = resp.followers

    dispatch(dismissError())
    // dispatch(setAwaitingPasswordChange(true))
    dispatch(saveUserID(userID))
    // setUserContext(userID)
    // dispatch(getFCMToken())
    // dispatch(startListeningFCMTokenRefresh())
    // dispatch(setDistributionOnServer())
    dispatch(setDoingInitialLoad(true))
    return PouchCouch.localReplicateDB('all', [...following, userID], followers)
  }).then(() => {
    return PouchCouch.localLoad()
  }).then(localData => {
    dispatch(localDataLoaded(localData))
    dispatch(setDoingInitialLoad(false))
    dispatch(switchRoot(FEED))
    dispatch(startListeningFCM())
  }).catch(e => {
    logDebug(JSON.stringify(e), 'this error specifically')
    dispatch(errorOccurred(e.message))
  })
}

export function tryToLoadStateFromDisk (dispatch) {
  return Promise.all([
    LocalStorage.loadLocalState(),
    LocalStorage.loadCurrentRideState()
  ]).then(([localState, currentRideState]) => {
    if (localState) {
      dispatch(loadLocalState(localState))
    } else {
      logInfo('no cached local state found')
    }

    if (currentRideState) {
      dispatch(loadCurrentRideState(currentRideState))
    } else {
      logInfo('no cached current ride state found')
    }
  })
}

export function configureBackgroundGeolocation () {
  return new Promise((res, rej) => {
    BackgroundGeolocation.configure(
      {
        desiredAccuracy: BackgroundGeolocation.HIGH_ACCURACY,
        locationProvider: BackgroundGeolocation.RAW_PROVIDER,
        distanceFilter: 0,
        maxLocations: 10,
        interval: 0,
        notificationTitle: 'You\'re out on a ride.',
        notificationText: 'Tap here to see your progress.',
      },
      () => {
        logInfo('Background geolocation configured')
        res()
      },
      (e) => {
        rej(e)
      },
    )
  })
}

export function stopListeningFCM () {
  // maybe delete token on server here?
  return firebase.iid().deleteToken('373350399276', 'GCM').then(() => {
    firebase.messaging().onTokenRefresh(() => {})
  })
}
