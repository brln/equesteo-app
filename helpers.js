const toRad = (deg) => {
  return deg * Math.PI / 180;
}

const toDeg = (rad) => {
  return rad * 180 / Math.PI
}


export const haversine = (lat1, lon1, lat2, lon2) => {
  const R = 3959; // mi
  const x1 = lat2 - lat1
  const dLat = toRad(x1)
  const x2 = lon2 - lon1
  const dLon = toRad(x2)
  const a = Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(dLon/2) * Math.sin(dLon/2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export const bearing = (lat1,lng1,lat2,lng2) => {
  let dLon = toRad(lng2-lng1);
  let y = Math.sin(dLon) * Math.cos(toRad(lat2));
  let x = Math.cos(toRad(lat1))*Math.sin(toRad(lat2)) - Math.sin(toRad(lat1))*Math.cos(toRad(lat2))*Math.cos(dLon);
  let brng = toDeg(Math.atan2(y, x));
  return ((brng + 360) % 360);
}

export const rideCoordsToMapCoords = (rideCoords) => {
  const sorted = [...rideCoords].sort((a, b) => {
    return new Date(a.timestamp) - new Date(b.timestamp);
  })

  return sorted.map((apiCoord) => {
    return {
      latitude: parseFloat(apiCoord.latitude),
      longitude: parseFloat(apiCoord.longitude),
    }
  })
}

export const unixTimeNow = () => {
  return Math.floor(new Date().getTime())
}

export const profilePhotoURL = (id) => {
  return `https://s3.amazonaws.com/equesteo-profile-photos/full_size/${id}.jpeg`
}

// from https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
export function generateUUID () { // Public Domain/MIT
    let d = new Date().getTime();
    if (typeof performance !== 'undefined' && typeof performance.now === 'function'){
        d += performance.now(); //use high-precision timer if available
    }
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        var r = (d + Math.random() * 16) % 16 | 0;
        d = Math.floor(d / 16);
        return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
    });
}

function urlParams (params) {
  return Object.keys(params).map(k => k + '=' + encodeURIComponent(params[k])).join('&')
}

export function staticMap (ride) {
  const STATIC_MAPS_API_KEY = ' AIzaSyBhUmpq-7uQ2JaqtrHO3hpfeFHynVpo8xQ'
  const ROOT_URL = 'https://maps.googleapis.com/maps/api/staticmap?'
  const queryStringParams = {
      size: '400x400',
      format: 'png',
      maptype: 'terrain',
  }
  const pathStyle = 'color:0xff0000ff|weight:5'
  let pathCoords = ''
  for (let coordinate of ride.rideCoordinates) {
    pathCoords += `|${coordinate.latitude},${coordinate.longitude}`
  }

  queryStringParams['path'] = pathStyle + pathCoords
  queryStringParams['key'] = STATIC_MAPS_API_KEY
  const queryString = urlParams(queryStringParams)
  return ROOT_URL + queryString
}
