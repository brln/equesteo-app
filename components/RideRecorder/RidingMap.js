import React, { Component } from 'react';
import MapView from 'react-native-maps';

import {
  Image,
  StyleSheet,
  Text,
  View
} from 'react-native';

import { bearing, rideCoordsToMapCoords } from '../../helpers'

export default class RidingMap extends Component {
  constructor (props) {
    super(props)
    this.state = {}
    this.fitToElements = this.fitToElements.bind(this)
    this.changeBearing = this.changeBearing.bind(this)
    this.fitToElements = this.fitToElements.bind(this)
    this.gpsStatusImage = this.gpsStatusImage.bind(this)
  }

  shouldComponentUpdate(nextProps) {
    let should = false
    if (nextProps.rideCoords.length !== this.props.rideCoords.length) {
      should = true
    }
    return should
  }

  componentDidUpdate(prevProps) {
    if (prevProps.rideCoords.length === 0 && this.props.rideCoords.length === 1) {
      this.map.animateToRegion(this.fitToElements())
    }
    this.changeBearing()
  }

  fitToElements() {
    const lats = this.props.rideCoords.map(c => c.latitude)
    const longs = this.props.rideCoords.map(c => c.longitude)
    const maxLat = Math.max(...lats)
    const minLat = Math.min(...lats)
    const maxLong = Math.max(...longs)
    const minLong = Math.min(...longs)
    const latitude = ((maxLat - minLat) / 2) + minLat || 0
    const longitude = ((maxLong - minLong) / 2) + minLong || 0
    const latitudeDelta = Math.max((maxLat - minLat) * 1.2, 0.01)
    const longitudeDelta = Math.max((maxLong - minLong) * 1.2, 0.01)
    return {
      latitude,
      longitude,
      latitudeDelta,
      longitudeDelta,
    }

  }

  changeBearing () {
    let newBearing = 0
    let coordinates = this.props.rideCoords
    const lastCoord = coordinates[coordinates.length -1]
    if (coordinates.length > 1) {
      newBearing = bearing(
        coordinates[coordinates.length - 2].latitude,
        coordinates[coordinates.length - 2].longitude,
        lastCoord.latitude,
        lastCoord.longitude
      )
      this.map.animateToNavigation(lastCoord, newBearing, 45)
    }
  }

  gpsStatusImage () {
    if (this.props.lastLocation) {
      switch (true) {
        case this.props.lastLocation.accuracy < 10:
          return require('../../img/gps/gps5.png')
        case this.props.lastLocation.accuracy < 15:
          return require('../../img/gps/gps4.png')
        case this.props.lastLocation.accuracy < 20:
          return require('../../img/gps/gps3.png')
        case this.props.lastLocation.accuracy < 25:
          return require('../../img/gps/gps2.png')
        default:
          return require('../../img/gps/gps1.png')
      }
    } else {
      return null
    }
  }

  render() {
    return (
      <View style ={styles.container}>
        <View style={{flex: 1}}>
          <MapView
            style={styles.map}
            ref={ref => this.map = ref}
            initialRegion={this.fitToElements()}
          >
            <MapView.Polyline
              style={styles.map}
              coordinates={this.props.rideCoords}
              strokeColor="#dc0202"
              strokeWidth={5}
            >
            </MapView.Polyline>
          </MapView>
        </View>
        <View style={{position: 'absolute', right: 0, top: 0}} >
          <Image source={this.gpsStatusImage()} style={{width: 50, height: 50}}/>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  map: {
    ...StyleSheet.absoluteFillObject,
  },
});

