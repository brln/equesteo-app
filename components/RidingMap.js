import React, { Component } from 'react';
import MapView from 'react-native-maps';

import {
  StyleSheet,
  View
} from 'react-native';

import { bearing } from '../helpers'


export default class RidingMap extends Component {
  constructor (props) {
    super(props)
    this.state = {}
    this.fitToElements = this.fitToElements.bind(this)
    // this.changeBearing = this.changeBearing.bind(this)
    this.fitToElements = this.fitToElements.bind(this)
  }

  fitToElements() {
    const maxLat = Math.max(...this.props.rideCoords.map((c) => c.latitude))
    const minLat = Math.min(...this.props.rideCoords.map((c) => c.latitude))
    const maxLong = Math.max(...this.props.rideCoords.map((c) => c.longitude))
    const minLong = Math.min(...this.props.rideCoords.map((c) => c.longitude))
    debugger
    const latitude = ((maxLat - minLat) / 2) + minLat || 0
    const longitude = ((maxLong - minLong) / 2) + minLong || 0
    const latitudeDelta = Math.max((maxLat - minLat) * 1.2, 0.01)
    const longitudeDelta = Math.max((maxLong - minLong) * 1.2, 0.01)
    return {
      latitude,
      longitude,
      latitudeDelta,
      longitudeDelta,
    }

  }

  changeBearing () {
    let newBearing = 0
    let coordinates = this.props.rideCoords
    if (coordinates.length > 1) {
      newBearing = bearing(
        coordinates[coordinates.length - 2].latitude,
        coordinates[coordinates.length - 2].longitude,
        coordinates[coordinates.length - 1].latitude,
        coordinates[coordinates.length - 1].longitude
      )
      this.map.animateToBearing(newBearing, 500)
    }
  }

  render() {
    return (
      <View style ={styles.container}>
        <MapView
          style={styles.map}
          ref={ref => this.map = ref}
          region={this.fitToElements()}
          // onRegionChangeComplete={this.changeBearing}
        >
          <MapView.Polyline
            style={styles.map}
            coordinates={this.props.rideCoords}
            strokeColor="#dc0202"
            strokeWidth={5}
          >
          </MapView.Polyline>
        </MapView>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  map: {
    ...StyleSheet.absoluteFillObject,
  },
});

