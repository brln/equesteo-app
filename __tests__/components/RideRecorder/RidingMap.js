import { fromJS, Map } from 'immutable'
import React from 'react';
import renderer from 'react-test-renderer';

import { unixTimeNow } from '../../../helpers'
import RidingMap from '../../../components/RideRecorder/RidingMap'

test('renders', () => {
  const rideCoords = fromJS({rideCoordinates: [[1, 1, 1, 1]]})
  const mapRegionChanged = () => {}
  const recenter = () => {}
  const tree = renderer.create(
    <RidingMap
      currentRideCoordinates={rideCoords}
      heading={1}
      centerCoordinate={[1, 1]}
      lastLocation={Map({latitude: 1, longitude: 1})}
      mapRegionChanged={mapRegionChanged}
      recenter={recenter}
      refiningLocation={Map({latitude: 1, longitude: 1})}
      setMapRef={() => {}}
      userControlledMap={false}
      zoomLevel={2}
    />).toJSON();
  expect(tree).toMatchSnapshot()
});
