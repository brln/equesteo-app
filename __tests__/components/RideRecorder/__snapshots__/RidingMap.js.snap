// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "left": 10,
        "position": "absolute",
        "top": 10,
        "width": 46,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 4,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "padding": 3,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "height": 40,
                "marginRight": 5,
                "width": 40,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
